libs/wire-api/src/Wire/API/MLS/Serialisation.hs

```diff
@@ -121,10 +128,51 @@ decodeMLS' = decodeMLS . LBS.fromStrict
 -- Return an error message in case of failure.
 decodeMLSWith :: Get a -> LByteString -> Either Text a
 decodeMLSWith p b = case runGetOrFail p b of
-  Left (_, _, msg) -> Left (T.pack msg)
+  Left (_, _, msg) -> Left (Text.pack msg)
   Right (remainder, pos, x)
     | LBS.null remainder -> Right x
-    | otherwise -> Left $ "Trailing data at position " <> T.pack (show pos)
+    | otherwise -> Left $ "Trailing data at position " <> Text.pack (show pos)
 
 decodeMLSWith' :: Get a -> ByteString -> Either Text a
 decodeMLSWith' p = decodeMLSWith p . LBS.fromStrict
+
+-- | An MLS value together with its serialisation.
+--
+-- This can be used whenever we need to parse an object, but at the same time
+-- retain the original serialised bytes (e.g. for signature verification, or to
+-- forward them verbatim).
+data RawMLS a = RawMLS
+  { rmRaw :: ByteString,
+    rmValue :: a
+  }
+  deriving stock (Eq, Show, Functor, Foldable, Traversable)
```

Comment:
What does it mean to have a functor instance for this type? When we map over and change `rmValue`, `rmRaw` stays the same, but that `rmRaw` has nothing to do with the new `rmValue`. Is that OK?

Reply:
Well, technically it is a valid functor and comonad, but you're right that it might be slightly misleading. I guess there is an unwritten assumption here that the `rmRaw` field is somehow the serialised form of the corresponding `rmValue`, but of course that's not enforced in any sense. To avoid any confusion, I removed the `Comonad` instance in 3b8a62e61d3d7c947d42b341d4330f4a1aa8fafe.